Summer ospp project 210370190 mid-term report (#449)
JuliaReinforcementLearning/ReinforcementLearning.jl: docs/homepage/blog/ospp_mid-term_report_210370190/index.md

    update!(agent.trajectory, agent.policy, env, stage)
    update!(agent.policy, agent.trajectory, env, stage)
end

function (agent::Agent)(stage::PreActStage, env::AbstractEnv, action)
    update!(agent.trajectory, agent.policy, env, stage, action)
    update!(agent.policy, agent.trajectory, env, stage)
end
```

And when running the experiment, based on the built-in [`run`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/639717388fb41199c98b90406bea76232bc6294d/src/ReinforcementLearningCore/src/core/run.jl#L16) function, the `agent` can update its policy and trajectory based on the behaviors that we have defined. Thanks to the `multiple dispatch` in Julia,  the **main focus** when implementing a new algorithm is how to **customize the behavior** of collecting the training information and updating the policy when at the specific stage. For more details, you can refer to this [blog](https://juliareinforcementlearning.org/blog/an_introduction_to_reinforcement_learning_jl_design_implementations_thoughts/#21_the_general_workflow).

### 2.2 Neural Fictitious Self-play(NFSP) algorithm

#### Brief Introduction

> Neural Fictitious Self-play(NFSP)\dcite{DBLP:journals/corr/HeinrichS16} algorithm is a useful multi-agent algorithm that works well for imperfect-information games. Each agent who applies the `NFSP` algorithm will include one `Reinforcement Learning`(RL) agent and one `Supervised Learning`(SL) agent. **RL agent** works to find the best response to the state from the self-play process, and **SL agent** works to learn the best response from RL agent's policy. What's more, `NFSP` also uses two technical innovations to ensure stability, including [reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling) for SL agent and anticipatory dynamics\dcite{1406126} when training.

Reviewer: Better to add a picture here to explain this algorithm clearly.

#### Implementation

In RLZoo.jl, I implement the [`NFSPAgent`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningZoo/src/algorithms/nfsp/nfsp.jl) which define the `NFSPAgent` struct and design its behaviors according to the `NFSP` algorithm\dcite{DBLP:journals/corr/HeinrichS16}, including collecting needed information and how to update the policy. And the [`NFSPAgentManager`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningZoo/src/algorithms/nfsp/nfsp_manager.jl) is a special multi-agent manager that all agents apply `NFSP` algorithm. Besides, the [`abstract_nfsp`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningZoo/src/algorithms/nfsp/abstract_nfsp.jl) customize the `run` function for `NFSPAgentManager`.

> Since the core of the algorithm is how to customize the `NFSPAgent`, the following content in this section will only be around it. The structure of `NFSPAgent` is as the following:

Reviewer: ```suggestion
Since the core of the algorithm is how to customize the `NFSPAgent`, the following content in this section will only be around it. The structure of `NFSPAgent` is as the following:
```
```suggestion
Since the core of the algorithm is to define the behavior of the `NFSPAgent`, I'll explain how it is done as the following:
```

```Julia
mutable struct NFSPAgent <: AbstractPolicy
    rl_agent::Agent
    sl_agent::Agent
    η # anticipatory parameter
    rng
    update_freq::Int # update frequency
    update_step::Int # count the step
    mode::Bool # `true` for best response mode(RL agent's policy), `false` for  average policy mode(SL agent's policy). Only used in training.
end
```

> Based on 2.1, the core of the `NFSPAgent` is customized behaviors on the specific stage:

Reviewer: ```suggestion
Based on our discussion in section 2.1, the core of the `NFSPAgent` is to customize its behavior in different stages:
```

- PreEpisodeStage

> Here, `NFSPAgent` should set train mode based on the anticipatory dynamics\dcite{1406126} and delete the terminated state and dummy action if having gone through one episode before. Note that here deleting the terminated state and dummy action is necessary for the algo(see the [note](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/4e5d258798088b1c628401b6b9de18aa8cbb3ab3/src/ReinforcementLearningCore/src/policies/agents/agent.jl#L134)), otherwise may occur to have some unreliable samples.

Reviewer: ```suggestion
Here, the `NFSPAgent` should be set to the training mode based on the anticipatory  dynamics\dcite{1406126}。 Besides, the terminated state and dummy action of the last episode must be removed at the beginning of each episode. (see also the [note](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/4e5d258798088b1c628401b6b9de18aa8cbb3ab3/src/ReinforcementLearningCore/src/policies/agents/agent.jl#L134))
```

```Julia
function (π::NFSPAgent)(stage::PreEpisodeStage, env::AbstractEnv, ::Any)
    # delete the terminal state and dummy action.
    update!(π.rl_agent.trajectory, π.rl_agent.policy, env, stage)

    # set the train's mode before the episode.(anticipatory dynamics)
    π.mode = rand(π.rng) < π.η
end
```

- PreActStage

> In this stage, `NFSPAgent` should collect the personal information `state` and `action` to the RL agent's trajectory, and if on the `best response mode`, also update the SL agent's trajectory. Besides, if satisfying the condition of updating, here also need to update the inner agents. The code is just like the following:

Reviewer: ```suggestion
In this stage, the `NFSPAgent` should collect the personal information of `state` and `action` and add them into the **RL** agent's trajectory. If it is set to the `best response mode`, we also need to update the **SL** agent's trajectory. Besides, if the condition of updating is satisfied, the inner agents also need to be updated. The code is just like the following:
```

```Julia
function (π::NFSPAgent)(stage::PreActStage, env::AbstractEnv, action)
    rl = π.rl_agent
    sl = π.sl_agent
    # update trajectory
    if π.mode
        update!(sl.trajectory, sl.policy, env, stage, action)
        rl(stage, env, action)
    else
        update!(rl.trajectory, rl.policy, env, stage, action)
    end

    # update policy
    π.update_step += 1
    if π.update_step % π.update_freq == 0
        if π.mode
            update!(sl.policy, sl.trajectory)
        else
            rl_learn!(rl.policy, rl.trajectory)
            update!(sl.policy, sl.trajectory)
        end
    end
end
```

- PostActStage

> Here, the agent needs to collect the personal `reward` and  the `is_terminated` judgment of the current state to the RL agent's trajectory.

Reviewer: ```suggestion
Here, the agent needs to add the personal `reward` and  the `is_terminated` result of the current state into the **RL** agent's trajectory.
```

```Julia
function (π::NFSPAgent)(::PostActStage, env::AbstractEnv, player::Any)
    push!(π.rl_agent.trajectory[:reward], reward(env, player))
    push!(π.rl_agent.trajectory[:terminal], is_terminated(env))
end
```

- PostEpisodeStage

> When one episode is terminated, the agent should collect the terminated state and a dummy action(see the [note](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/4e5d258798088b1c628401b6b9de18aa8cbb3ab3/src/ReinforcementLearningCore/src/policies/agents/agent.jl#L134)) to the RL agent's trajectory. Also, the reward and terminal judgment need to be corrected to avoid getting wrong samples when playing the sequential or terminal_reward games.

Reviewer: ```suggestion
When one episode is terminated, the agent should push the terminated state and a dummy action(see the [note](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/4e5d258798088b1c628401b6b9de18aa8cbb3ab3/src/ReinforcementLearningCore/src/policies/agents/agent.jl#L134)) into the **RL** agent's trajectory. Also, the reward and terminal judgment need to be corrected to avoid getting wrong samples when playing the sequential or terminal_reward games.
```

```Julia
function (π::NFSPAgent)(::PostEpisodeStage, env::AbstractEnv, player::Any)
    rl = π.rl_agent
    sl = π.sl_agent
    # update trajectory
    if !rl.trajectory[:terminal][end]
        rl.trajectory[:reward][end] = reward(env, player)
        rl.trajectory[:terminal][end] = is_terminated(env)
    end

    action = rand(action_space(env, player))
    push!(rl.trajectory[:state], state(env, player))
    push!(rl.trajectory[:action], action)
    if haskey(rl.trajectory, :legal_actions_mask)
        push!(rl.trajectory[:legal_actions_mask], legal_action_space_mask(env, player))
    end
    
    # update the policy    
    ...
end
```

#### Usage

> According to the paper\dcite{DBLP:journals/corr/HeinrichS16}, here, the RL agent is default as `QBasedPolicy` with `CircularArraySARTTrajectory.` The SL agent is default as `BehaviorCloningPolicy` with `ReservoirTrajectory.` So you can customize the agent under the restriction and test the algo on any interested multi-agent game. **Note that** if the game's states can't be used as the network's input, you need to [wrap](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/50756bbe9e1925a9320d1abdbbc6255c1b4a27f1/src/ReinforcementLearningEnvironments/src/environments/wrappers/StateTransformedEnv.jl#L9) them before using the algorithm.

Reviewer: ```suggestion
According to the paper\dcite{DBLP:journals/corr/HeinrichS16}, by default the RL agent is as `QBasedPolicy` with `CircularArraySARTTrajectory.` The SL agent is default as `BehaviorCloningPolicy` with `ReservoirTrajectory.` So you can customize the agent under the restriction and test the algorithm on any interested multi-agent game. **Note that** if the game's states can't be used as the network's input, you need to add a [wrapper](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/50756bbe9e1925a9320d1abdbbc6255c1b4a27f1/src/ReinforcementLearningEnvironments/src/environments/wrappers/StateTransformedEnv.jl#L9) to the environment before applying the algorithm.
```

\dfig{body;JuliaRL_NFSP_KuhnPoker.png;Result of the experiment.}

### 2.3 Multi-agent Deep Deterministic Policy Gradient(MADDPG) algorithm

#### Brief Introduction

The Multi-agent Deep Deterministic Policy Gradient(MADDPG)\dcite{DBLP:journals/corr/LoweWTHAM17} algorithm improves the [Deep Deterministic Policy Gradient(DDPG)](https://spinningup.openai.com/en/latest/algorithms/ddpg.html), which works well on multi-agent games. Based on the DDPG, the critic of each agent in MADDPG can get all agents' policies according to the paper's hypothesis\dcite{DBLP:journals/corr/LoweWTHAM17}, including their personal states and actions, which can help get a more reasonable score of the actor's policy.

#### Implementation

> Since there has been [`DDPGPolicy`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/50756bbe9e1925a9320d1abdbbc6255c1b4a27f1/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/ddpg.jl#L47) in the RLZoo, I implement the [`MADDPGManager`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl) which is a special multi-agent manager that all agents apply `DDPGPolicy` with one **improved critic**. The structure of `MADDPGManager` is as the following:

Reviewer: ```suggestion
Given that the [`DDPGPolicy`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/50756bbe9e1925a9320d1abdbbc6255c1b4a27f1/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/ddpg.jl#L47) is already implemented in the RLZoo, I implement the [`MADDPGManager`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl) which is a special multi-agent manager that all agents apply `DDPGPolicy` with one **improved critic**. The structure of `MADDPGManager` is as the following:
```

Author: Thanks for your patience in reviewing and sorry for my terrible report. 😞 I'll update my report follow your advice soon!
